ls1intum/Artemis
# `Programming exercises`: Reduce the network traffic in the submission dashboard

### Checklist
#### General


#### Server

### Motivation and Context
Following issue #4741, when the submission dashboard is opened, a large JSON is exchanged between server and client, which has a lot of unnecessary and duplicate data.
For each student submission, a copy of the `exercise` object (including the problem statement) and the `course`-object is sent to the client. To get the number of Submissions, the whole `Set<Submission>` is send to the client to determine the size. (See issue #4741 , the JSON for ≈1100 participants has ≈43MB)_
This leads to long waiting and processing times when the submission dashboard is opened in the client. 
This PR removes the data in the server and thus greatly reduces the size of the transferred JSON file.
```
Current structure of the JSON (partly)
ProgrammingSubmission
-> participation
     -> exercise
          -> course
     -> submissions[]
```

### Description
Two methods in the `ProgrammingSubmissionService` are used to retrieve data for the REST-Call:
- `getProgrammingSubmissions()`for the submission dashboard
- `getAllProgrammingSubmissionsAssessedByTutorForCorrectionRoundAndExercise()` for the assessment dashboard of a programming exercise 
A private method `removeExerciseAndSubmissionSet()` has been added, setting attribute value for `exercise` to null. If the `boolean removeSubmissionSet` is true, also the `Set<Submission>` will be set to `null` and the number of submissions is instead stored in the variable `submissionCount` of the class `Participation`.
_(If the data is retrieved from the `studentParticipationRepository`, the `Set<Submission>` is appended and thus should be removed. If the data is retrieved from `submissionRepository`, the `Set<Submission>` is not appended)_

In the client, only the number of submissions must be changed to the new attribute and away from `submissions.length`
```
New structure of the JSON (partly)
ProgrammingSubmission
-> participation
     -> submissionsCount
``` 

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise with submissions and selected manual assessment.
- 1 Exam with a Programming Exercise with submissions and selected manual assessment.

1. Log in to Artemis
2. Navigate to Course Administration -> Exercises -> (Programming Exercise) -> Submission 
3. Check if the submission dashboard is displayed correctly and the transferred JSON does not have an `exercise` attribute under `participation` (Note: `course` is an attribute of `exercise`)
4. Navigate to the Assessment Dashboard and assess the Programming Exercise
5. Assess at least one submission of the programming exercise
6.  Check if the assessment dashboard for the programming exercise correctly displays the assessed submissions by the tutor-account and the transferred JSON does not have an `exercise` attribute under `participation`  (Note: the corresponding method to display the assessment per tutor uses the same REST-Call)
7. Do the same in the Exam-Mode

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage






### Screenshots
